File: minigit-5/ts/src/commands/log.ts
// log command - Show commit history

import { findRepoRoot, formatDate, parseTimestamp } from '../utils';
import { getCommit, readObject, getBlob, getTreeFromTreeIsh, walkTree } from '../objects';
import { getHeadCommit, resolveRevision, listBranches } from '../refs';
import { generateDiff, formatDiff, formatDiffStat } from '../diff-algo';
import { FileDiff } from '../types';

export function log(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  // Parse arguments
  let oneline = false;
  let showAll = false;
  let graph = false;
  let showStat = false;
  let limit = -1;
  let startRevision: string | null = null;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--oneline') {
      oneline = true;
    } else if (arg === '--all') {
      showAll = true;
    } else if (arg === '--graph') {
      graph = true;
    } else if (arg === '--stat') {
      showStat = true;
    } else if (arg === '-n' && i + 1 < args.length) {
      limit = parseInt(args[i + 1], 10);
      i++;
    } else if (!arg.startsWith('-')) {
      startRevision = arg;
    }
  }

  // Collect starting commits
  const startShas: string[] = [];

  if (showAll) {
    // Get all branch heads
    const branches = listBranches(repoRoot);
    for (const branch of branches) {
      try {
        const sha = resolveRevision(branch, repoRoot);
        if (!startShas.includes(sha)) {
          startShas.push(sha);
        }
      } catch {
        // Skip invalid branches
      }
    }
    // Also include HEAD
    const headSha = getHeadCommit(repoRoot);
    if (headSha && !startShas.includes(headSha)) {
      startShas.push(headSha);
    }
  } else if (startRevision) {
    try {
      const sha = resolveRevision(startRevision, repoRoot);
      startShas.push(sha);
    } catch (e: any) {
      console.error(`fatal: ${e.message}`);
      return 1;
    }
  } else {
    const headSha = getHeadCommit(repoRoot);
    if (!headSha) {
      console.error('fatal: your current branch does not have any commits yet');
      return 1;
    }
    startShas.push(headSha);
  }

  // Walk commits in reverse chronological order
  const visited = new Set<string>();
  const queue: string[] = [...startShas];
  const commits: Array<{ sha: string; date: Date }> = [];

  while (queue.length > 0) {
    const sha = queue.shift()!;
    if (visited.has(sha)) continue;
    visited.add(sha);

    try {
      const commit = getCommit(sha, repoRoot);

      // Parse date from author line
      const authorParts = commit.author.split(' ');
      const timestamp = parseInt(authorParts[authorParts.length - 2], 10);
      const date = new Date(timestamp * 1000);

      commits.push({ sha, date });

      // Add parents to queue
      for (const parent of commit.parents) {
        if (!visited.has(parent)) {
          queue.push(parent);
        }
      }
    } catch {
      // Skip invalid commits
    }
  }

  // Sort by date (newest first)
  commits.sort((a, b) => b.date.getTime() - a.date.getTime());

  // Apply limit
  const toShow = limit > 0 ? commits.slice(0, limit) : commits;

  // Output
  for (let i = 0; i < toShow.length; i++) {
    const { sha } = toShow[i];
    const commit = getCommit(sha, repoRoot);

    if (oneline) {
      const prefix = graph ? '* ' : '';
      const shortSha = sha.slice(0, 7);
      const firstLine = commit.message.split('\n')[0];
      console.log(`${prefix}${shortSha} ${firstLine}`);
    } else {
      if (graph && i > 0) {
        console.log('|');
      }
      const prefix = graph ? '* ' : '';
      console.log(`${prefix}commit ${sha}`);

      // Parse author
      const authorMatch = commit.author.match(/^(.+) <(.+)> (\d+) ([+-]\d{4})$/);
      if (authorMatch) {
        console.log(`Author: ${authorMatch[1]} <${authorMatch[2]}>`);
        const date = new Date(parseInt(authorMatch[3], 10) * 1000);
        console.log(`Date:   ${formatDate(date)}`);
      } else {
        console.log(`Author: ${commit.author}`);
      }

      console.log('');
      const messageLines = commit.message.split('\n');
      for (const line of messageLines) {
        console.log(`    ${line}`);
      }

      if (showStat) {
        // Show stat
        const diffs = getCommitDiffs(sha, commit, repoRoot);
        if (diffs.length > 0) {
          console.log('');
          process.stdout.write(formatDiffStat(diffs));
        }
      }

      if (i < toShow.length - 1) {
        console.log('');
      }
    }
  }

  return 0;
}

function getCommitDiffs(sha: string, commit: ReturnType<typeof getCommit>, repoRoot: string): FileDiff[] {
  const diffs: FileDiff[] = [];
  const currentTree = commit.tree;
  const currentFiles = walkTree(currentTree, '', repoRoot);

  if (commit.parents.length > 0) {
    const parentSha = commit.parents[0];
    const parentTree = getTreeFromTreeIsh(parentSha, repoRoot);
    const parentFiles = walkTree(parentTree, '', repoRoot);

    const allPaths = new Set([...parentFiles.keys(), ...currentFiles.keys()]);

    for (const filePath of allPaths) {
      const parentEntry = parentFiles.get(filePath);
      const currentEntry = currentFiles.get(filePath);

      if (!parentEntry && currentEntry) {
        const newContent = getBlob(currentEntry.sha, repoRoot).toString();
        const diff = generateDiff('', newContent, '/dev/null', filePath);
        diffs.push({ ...diff, oldPath: '/dev/null', newPath: filePath });
      } else if (parentEntry && !currentEntry) {
        const oldContent = getBlob(parentEntry.sha, repoRoot).toString();
        const diff = generateDiff(oldContent, '', filePath, '/dev/null');
        diffs.push({ ...diff, oldPath: filePath, newPath: '/dev/null' });
      } else if (parentEntry && currentEntry && parentEntry.sha !== currentEntry.sha) {
        const oldContent = getBlob(parentEntry.sha, repoRoot).toString();
        const newContent = getBlob(currentEntry.sha, repoRoot).toString();
        const diff = generateDiff(oldContent, newContent, filePath, filePath);
        if (diff.hunks.length > 0) {
          diffs.push(diff);
        }
      }
    }
  } else {
    // Initial commit
    for (const [filePath, entry] of currentFiles) {
      const content = getBlob(entry.sha, repoRoot).toString();
      const diff = generateDiff('', content, '/dev/null', filePath);
      diffs.push({ ...diff, oldPath: '/dev/null', newPath: filePath });
    }
  }

  return diffs;
}

File: minigit-5/py/commands/log.py
"""log command - Show commit history."""

import sys
from pathlib import Path
from datetime import datetime, timezone

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from objects import Commit
from refs import resolve_head, resolve_revision, list_branches, read_ref


def run(args: list[str]) -> int:
    """Show commit history."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    oneline = False
    show_all = False
    limit = None
    graph = False
    stat = False
    revision = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '--oneline':
            oneline = True
        elif arg == '--all':
            show_all = True
        elif arg == '--graph':
            graph = True
        elif arg == '--stat':
            stat = True
        elif arg == '-n' and i + 1 < len(args):
            limit = int(args[i + 1])
            i += 1
        elif not arg.startswith('-'):
            revision = arg
        i += 1

    # Determine starting points
    if show_all:
        start_shas = get_all_branch_heads(repo_root)
    elif revision:
        sha = resolve_revision(repo_root, revision)
        if sha is None:
            print(f"fatal: unknown revision '{revision}'", file=sys.stderr)
            return 1
        start_shas = [sha]
    else:
        head_sha = resolve_head(repo_root)
        if head_sha is None:
            print("fatal: your current branch does not have any commits yet", file=sys.stderr)
            return 1
        start_shas = [head_sha]

    # Collect commits
    commits = collect_commits(repo_root, start_shas, limit)

    if not commits:
        return 0

    # Output
    for i, (sha, commit) in enumerate(commits):
        if oneline:
            print_oneline(sha, commit, graph and i > 0)
        else:
            print_full(sha, commit, stat, repo_root)
            if i < len(commits) - 1:
                print()

    return 0


def get_all_branch_heads(repo_root: Path) -> list:
    """Get SHA of all branch heads."""
    shas = []
    for branch in list_branches(repo_root):
        sha = read_ref(repo_root, f'refs/heads/{branch}')
        if sha:
            shas.append(sha)
    return shas


def collect_commits(repo_root: Path, start_shas: list, limit: int = None) -> list:
    """Collect commits from starting points, sorted by date."""
    visited = set()
    commits = []
    queue = list(start_shas)

    while queue and (limit is None or len(commits) < limit):
        sha = queue.pop(0)
        if sha in visited:
            continue
        visited.add(sha)

        try:
            commit = Commit.read(sha, repo_root)
        except:
            continue

        commits.append((sha, commit))

        # Add parents to queue
        for parent in commit.parents:
            if parent not in visited:
                queue.append(parent)

    # Sort by timestamp (most recent first)
    commits.sort(key=lambda x: get_commit_timestamp(x[1]), reverse=True)

    if limit:
        commits = commits[:limit]

    return commits


def get_commit_timestamp(commit: Commit) -> int:
    """Extract timestamp from commit."""
    try:
        parts = commit.author_date.split()
        return int(parts[0])
    except:
        return 0


def print_oneline(sha: str, commit: Commit, show_graph: bool):
    """Print one-line commit format."""
    first_line = commit.message.split('\n')[0]
    prefix = "* " if show_graph else ""
    print(f"{prefix}{sha[:7]} {first_line}")


def print_full(sha: str, commit: Commit, stat: bool, repo_root: Path):
    """Print full commit format."""
    print(f"commit {sha}")

    # Show merge parents if any
    if len(commit.parents) > 1:
        parent_short = ' '.join(p[:7] for p in commit.parents)
        print(f"Merge: {parent_short}")

    print(f"Author: {commit.author} <{commit.author_email}>")
    print(f"Date:   {format_date(commit.author_date)}")
    print()

    # Indent message
    for line in commit.message.split('\n'):
        print(f"    {line}")

    if stat and commit.parents:
        print()
        print_stat(repo_root, commit.parents[0], sha)


def format_date(date_str: str) -> str:
    """Format git date string for display."""
    try:
        parts = date_str.split()
        ts = int(parts[0])
        tz = parts[1] if len(parts) > 1 else '+0000'

        # Parse timezone
        tz_hours = int(tz[:3])
        tz_mins = int(tz[0] + tz[3:5])

        dt = datetime.fromtimestamp(ts, tz=timezone.utc)
        return dt.strftime('%a %b %d %H:%M:%S %Y') + f" {tz}"
    except:
        return date_str


def print_stat(repo_root: Path, parent_sha: str, commit_sha: str):
    """Print stat for commit."""
    from objects import Tree

    try:
        parent_commit = Commit.read(parent_sha, repo_root)
        commit = Commit.read(commit_sha, repo_root)

        parent_files = get_all_files(repo_root, parent_commit.tree_sha)
        commit_files = get_all_files(repo_root, commit.tree_sha)

        # Find changes
        all_files = set(parent_files.keys()) | set(commit_files.keys())

        for name in sorted(all_files):
            if name not in parent_files:
                print(f" {name} | new file")
            elif name not in commit_files:
                print(f" {name} | deleted")
            elif parent_files[name] != commit_files[name]:
                print(f" {name} | modified")
    except:
        pass


def get_all_files(repo_root: Path, tree_sha: str, prefix: str = '') -> dict:
    """Get all files from a tree."""
    from objects import Tree

    files = {}
    tree = Tree.read(tree_sha, repo_root)

    for entry in tree.entries:
        full_path = f"{prefix}{entry.name}" if prefix else entry.name
        if entry.mode == 0o40000:
            files.update(get_all_files(repo_root, entry.sha, full_path + '/'))
        else:
            files[full_path] = entry.sha

    return files
